From 5d83324deaf5d8b8cad969c3b6edca49da1ccb57 Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Tue, 18 Feb 2025 22:14:58 +0100
Subject: [PATCH 10/10] cel-questone-2a: update platform_drivers with 6.11
 compatibility

platform_driver::remove() changed its return type to void, so we need to
provide different versions based on the kernel version.

Add variants using the following semantic patch:

<// <smpl>
@r@
identifier DRIVER;
identifier remove_fn;
fresh identifier remove_fn_wrap = remove_fn ## "_6_11";
position p;
@@
struct platform_driver DRIVER@p = {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+       .remove = remove_fn_wrap,
+#else
        .remove = remove_fn,
+#endif
};

@i depends on r@
@@
 #include <linux/version.h>
@depends on r && !i@
@@
 #include <...>
+#include <linux/version.h>
@depends on r@
identifier r.remove_fn;
identifier r.remove_fn_wrap;
@@
int remove_fn(...) {...}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+static void remove_fn_wrap(struct platform_device *pdev)
+{
+       remove_fn(pdev);
+}
+#endif
// </smpl>

applied with

    spatch --sp-file 6.11-platform_remove_sig.cocci --in-place --dir packages/platforms/celestica/x86-64/questone-2a/

Result taken as is, without any code style fixes.

Upstream-Status: Inappropriate [Questone 2a support not upstream]
Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 .../modules/builds/src/questone2a_baseboard_cpld.c    | 11 +++++++++++
 .../modules/builds/src/questone2a_switchboard.c       | 10 ++++++++++
 2 files changed, 21 insertions(+)

diff --git a/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_baseboard_cpld.c b/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_baseboard_cpld.c
index 1a5527f0d691..ae691b8367a0 100644
--- a/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_baseboard_cpld.c
+++ b/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_baseboard_cpld.c
@@ -28,6 +28,7 @@
 #include <linux/types.h>
 #include <uapi/linux/stat.h>
 #include <linux/string.h>
+#include <linux/version.h>
 
 #define DRIVER_NAME "sys_cpld"
 /**
@@ -377,10 +378,20 @@ static int baseboard_cpld_drv_remove(struct platform_device *pdev)
     sysfs_remove_group(&pdev->dev.kobj, &baseboard_cpld_attrs_grp);
     return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+static void baseboard_cpld_drv_remove_6_11(struct platform_device *pdev) {
+    baseboard_cpld_drv_remove(pdev);
+}
+#endif
 
 static struct platform_driver baseboard_cpld_drv = {
     .probe  = baseboard_cpld_drv_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+    .remove = __exit_p(baseboard_cpld_drv_remove_6_11),
+#else
+    
     .remove = __exit_p(baseboard_cpld_drv_remove),
+#endif
     .driver = {
         .name   = DRIVER_NAME,
     },
diff --git a/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_switchboard.c b/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_switchboard.c
index ac604adbd9be..765299723c6b 100644
--- a/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_switchboard.c
+++ b/packages/platforms/celestica/x86-64/questone-2a/modules/builds/src/questone2a_switchboard.c
@@ -1847,6 +1847,11 @@ static int seastone2_drv_remove(struct platform_device *pdev)
     devm_kfree(&pdev->dev, fpga_data);
     return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+static void seastone2_drv_remove_6_11(struct platform_device *pdev) {
+    seastone2_drv_remove(pdev);
+}
+#endif
 
 #ifdef TEST_MODE
     #define FPGA_PCI_BAR_NUM 2
@@ -1932,7 +1937,12 @@ static struct pci_driver pci_dev_ops = {
 
 static struct platform_driver seastone2_drv = {
     .probe  = seastone2_drv_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,11,0)
+    .remove = __exit_p(seastone2_drv_remove_6_11),
+#else
+    
     .remove = __exit_p(seastone2_drv_remove),
+#endif
     .driver = {
         .name   = DRIVER_NAME,
     },
-- 
2.51.0

